### ISSUE TITLE
Feature request: Like placeholder, but after promise is resolved

### ISSUE THREAD

**Issue by papas-source** [WebReflection/hyperHTML#85]
The addition of the `placeholder` in case a promise is returned, is a very very useful addition and I really found it kind of liberating!

Now I can use a function name in the placeholder (that returns an empty string or even `null`) to (for example) show a gray overlay that covers the screen and in the center/middle of it a message saying "please wait" with a nice spinner...

But it would be very useful to have a similar way to hide this overlay - as soon as the promise is resolved!

I believe It is an easy and nice addition!

TIA

**Comment by joshgillies** [comment]
If I'm not mistaken, that's exactly the mechanism of `placeholder` already. Given async content, the placeholder value will be used, up until the async value is available. At which point your placeholder value will be replaced by the resolved content.

As an example check out this CodePen: https://codepen.io/joshgillies/pen/QMqJmY?editors=0110

**Comment by papas-source** [comment]
@joshgillies thanks for your reply.

As soon as the promise is resolved, the currently bound node transforms to either text, html etc. 
The BOUND NODE. But I am talking about a mechanism that would not be bound to the node where the promise/placeholder is attached. 

You can imagine this as a callback, much like the jQuery's `$.ajax` `beforeSend` / `complete` setting!

**Comment by joshgillies** [comment]
So you're goal is to transform some other arbitrary node based on the resolution of your async value?

**Comment by papas-source** [comment]
You can say that, although this is not exactly and/or only that. 

I want to be able to automate most tasks, and since HyperHTML lets me control what happens BEFORE the async part, it would be nice to let me control what happens AFTER the async part also.

Yes, I agree that manipulating other nodes is actually a side effect of the `placeholder` feature, but believe me, it is going to be used like that a lot. And it would be nice to also have the "complementary" part ;)

It really is very easy, actually see [this plunk](http://plnkr.co/edit/Xc2kCUDW7edfAqLwuNnm?p=preview) where I quickly/dirty added that feature with the `done` key

**Comment by papas-source** [comment]
BTW, it is very easy to implement as a transformer:
transformer name: promise

```
hyperHTML.define(
  'promise',
  function(o){
    o.before && typeof o.before === 'function' && o.before()
    return Promise.resolve(o.promise).then( (res) => {
      o.after && typeof o.after == 'function' && o.after();
      return res;
    })
  }
)

....

hyperHTML.bind(document.getElementById('content'))`
  ${{
    promise: {
      promise: new Promise(resolve => {
        setTimeout(resolve, 3000, {html: "Here is your <b>HTML</b> content!"});
      }),
      before: show_overlay,
      after: hide_overlay
    }
  }}
`;

```


I just thought it would be nice as an addition to the core since it doesn't add any overhead, nor in filesize or performance

**Comment by WebReflection** [comment]
I'm not sure I understand the use case, but I've simplified the plunk
http://plnkr.co/edit/vSgq80a5whTuMA6WoMQE?p=preview

> Now I can use a function name in the placeholder (that returns an empty string or even null) to (for example) show a gray overlay that covers the screen and in the center/middle of it a message saying "please wait" with a nice spinner...

Well, you can use any node as placeholder directly, no need to return undefined or null and show the node manually. The placeholder will be there 'till it'd be replaced with some other content, as shown in my plunk.

> the currently bound node transforms to either text, html etc. The BOUND NODE.

I think you misunderstood what's the placeholder about.

Let me try to explain, you have container and two content, p and div

```js
// given two random DOM nodes
const p = document.createElement('p');
const div = document.createElement('div');
p.textContent = 'this is going to be legend ...';
div.textContent = '... dary!';

// this is how you use a placeholder
hyperHTML.bind(document.body)`
  <h1>Hello World</h1>
  ${{
    any: new Promise(r => setTimeout(r, 2000, div)),
    placeholder: p
  }}`;
```
Here the [Code Pen](https://codepen.io/WebReflection/pen/mMBZGp?editors=0010) with the example.


> HyperHTML lets me control what happens BEFORE the async part, it would be nice to let me control what happens AFTER the async part also.

You can even return asynchronously another Promise or any other intent so you definitively have all the control you want AFTER.


> I agree that manipulating other nodes is actually a side effect of the placeholder feature

There is no side effect in using the placeholder ... or at least I don't see any.


> It really is very easy, actually see this plunk where I quickly/dirty added that feature with the done key
> I just thought it would be nice as an addition to the core since it doesn't add any overhead

Beside your non issue, to my understanding and my eyes, the point of having extendable intents in core is exactly to cover any possible use case and none of them will land in core *unless* these are impossible to add otherwise or these are really a common case.

I see your use case rather a misunderstanding on how and where hyperHTML operates.

If you think that comment node is where is your content, as example, you are wrong. The content is before such node, any content you want would be right before indeed.

Please let me know if there's anything else I can explain, thank you.

**Comment by papas-source** [comment]
I understand completely where the hyperHTML operates, but it seems like I cannot make you understand the use case I want to pinpoint, and this is probably caused by my bad english :(

Just think of the gmail, for example. When you type a search string and press enter, a yellow "please wait" appears at the upper top area. When the search finishes and the results are visible to the user, this yellow stripe disappears. Now, this yellow stripe could be anywhere on the DOM, for example right before the closing `</body>`. And this yellow stripe shows up not only on search, but in many other situations in the user experience with Gmail. 

This is somehow what I want to achieve. To automate such tasks. by using a function name in the `placeholder`, I can achieve the showing of this div, but It would be nice to have the complementary "hiding" technique so I can hide it when the promise resolves, so there would be no need to hide it programmaticaly everytime (for example) an ajax request finishes!

> There is no side effect in using the placeholder ... or at least I don't see any

What I ment was that using the `placeholder` the way I describe, is a side effect, actually a "side use case", an "exploit" actually, since this is not the intended use case for the `placeholder`. But if you really think about it, you will see that a lot of developers could exploit this feature to achieve what I describe!

> none of them will land in core unless these are impossible to add otherwise or these are really a common case

To my understanding, the whole `placeholder` think could be achieved in many other ways, and there was no need to include it in the core. ;) But It made it to the core mostly because it was present as `until` in lit-html.

Anyway, it was just a suggestion that I thought it would complement the lib and automate mostly ajax calls. I can really achieve what I want by using the transformer I posted on a previous reply!

Thank you all for your replies!

I am closing this for now but If you need any more info on this, please let me know and I wuld be glad to explain more - well, to the best of my abilities in english!

**Comment by WebReflection** [comment]
>  the whole placeholder think could be achieved in many other ways

nope, it uses internals impossible to reach otherwise and it gives you maximum flexibility as companion for any other use case.

I'd like to be sure you understand all the options you have here.

Here an example of a transformer that receives your loading thing.

```js
// genric thingy, could be already on the DOM too
const thingy = document.createElement('p');
thingy.style.cssText = `
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  text-align: center;
  padding-top: 100px;
  background: silver;
`;
thingy.textContent = 'loading';

// define your transformer
hyperHTML.define('myThing', function (placeholder) {
  return {
    any: new Promise(
      r => setTimeout(r, 2000, ['<p>All good!</p>'])
    ),
    placeholder
  };
});

// show thingy, then whatever
hyperHTML.bind(document.body)`
  <div>
    <strong>what up</strong>
    ${{myThing: thingy}}
  </div>
`;
```

The related [pen here](https://codepen.io/WebReflection/pen/WEXowG?editors=0010).

Not only you can return an intent with a placeholder, you can also pass along anything you want like:

  * the DOM node to show
  * the ajax request to make
  * the REST API to call
  * an object with all these info

Accordingly, it looks like you're better off with your own `hyperHTML.define('ajax', options => doStuff(options))` which is why I don't see this as something needed in core.


> But It made it to the core mostly because it was present as until in lit-html

The only thing I've realized and quite possibly published in a way easier, portable, and more elegant way, is the need to extend intents.

Placeholders were already possible via promises but the required automation behind the scene on resolution could not have been done outside the library because the registry is internal and it'll be kept as such to grant reliability (it will be also enhanced soon).

Now I'd like to ask you a question:

you have a deadly simple way to define whatever you need, including the ability to hack around `placeholder` and use a function to do something else, everything as transformer.

What is it that you don't like about this? 'cause if nothing, I think the current state of the library demonstrated it's easy to implement any case you can think of, so I'm quite pleased this was easy to solve in user-land.
